### PR TITLE
 Support unmapped fields in search 'fields' option

### LINKED ISSUE #63690: Support unmapped fields in search 'fields' option.
We recommend omitting a field from the mappings when a user wants to avoid the indexing overhead, but still retrieve its values. This is usually accomplished by using an `object` field with `enabled: false` or `dynamic: false`. Since 'fields' is meant to be the central place to retrieve document content, it should allow for loading these unmapped values.

Longer-term we'd like to discourage users from having unmapped fields, and instead introduce lightweight ways to map data without indexing it. But in the meantime, we think it's valuable to support unmapped fields.

One proposal is to add a flag `include_unmapped`:

```
{
  "query": { "match_all": {} },
  "fields": [{
    "field": "headers.*",
    "include_unmapped": true
  }]
}
```

When `include_unmapped` is specified, the 'fields' option would do the following:
1. Find all mapped leaf fields matching the patterns, collect + parse their values and add those to the hit. (This is what the option currently does).
2. Also find all leaf fields in _source matching the patterns. If we find any fields that were not included in 1, add those to the hit as well. The fields are returned as-is, ignoring the 'format’ option.

The flag could default to 'false', and eventually be phased out as it becomes less common to have unmapped fields.

Some questions to consider:
* Do we prefer adding a flag or just returning unmapped fields by default? It's always nice to avoid flags, and as we move towards a set-up where unmapped fields are rare, the distinction would become unimportant anyways. But the behavior is a bit hacky/ tricky, and I like that the flag indicates that it's 'non-standard' to be retrieving unmapped fields (in line with our longer term vision).
* Would returning unmapped fields by default hurt fetch performance in the 'happy case' where all fields are mapped?
* Is it fine to just return the unmapped values in a flattened list? This would drop structure in values like unmapped object arrays, ranges, etc.

### PR DESCRIPTION
**Pull request by cbuescher**
Currently, the 'fields' option only supports fetching mapped fields. Since
'fields' is meant to be the central place to retrieve document content, it
should allow for loading unmapped values. This change adds implementation and
tests for this feature.

Closes #63690

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/search/fetch/subphase/FieldFetcherTests.java
Comment (jtibshirani) [inline server/src/test/java/org/elasticsearch/search/fetch/subphase/FieldFetcherTests.java:643]: Small comment: there's getting to be a lot of convenience methods, maybe we could prefer just constructing and passing in `FieldAndFormat`?

File: server/src/test/java/org/elasticsearch/search/fetch/subphase/FieldFetcherTests.java
Comment (jtibshirani) [inline server/src/test/java/org/elasticsearch/search/fetch/subphase/FieldFetcherTests.java:400]: Nice test coverage! Some other cases that'd be good to check:
* Source documents that have dots in field names (like `unmapped_field.a: value`).
* Mixed dot and object notation. We recently found a bug in the 'fields' option related to this: https://github.com/elastic/elasticsearch/issues/65499.

**Review by jtibshirani — COMMENTED** [review]
The overall approach looks good to me, I left some detailed comments.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldAndFormat.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldAndFormat.java:90]: Sure, I'll look into Explicit, I personally really dislike using `null` as carrying semantic meaning like `unset`, which is my main motivation for Optional here.

File: server/src/test/java/org/elasticsearch/search/fetch/subphase/FieldFetcherTests.java
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/search/fetch/subphase/FieldFetcherTests.java:427]: Sorry, this was a left over, we already discussed this on the WIP. Will remove the TODO.

File: server/src/main/java/org/elasticsearch/action/search/SearchRequestBuilder.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/action/search/SearchRequestBuilder.java:324]: sure. Will try to reduce it to a convenience method and one using `FieldsAndFormat`

File: server/src/test/java/org/elasticsearch/search/fetch/subphase/FieldFetcherTests.java
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/search/fetch/subphase/FieldFetcherTests.java:611]: The previous test case 'testMappedFieldNotOverwritten' tests the special case where we don't want to return a mapped field that is not present in the document but in `_source` (e.g. because of ignore_malformed). Here I wanted to test various field pattern behaviours for object structures in an isolated test. So I think its not ideal to merge them.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java:136]: It took me a while to see how this could be an issue but I think you mean cases like:
```
{
	"user": [{
		"name": "eve",
		"scores": [1, 2, 3]
	}, {
		"name": "adam",
		"scores": [4, 5, 6]
	}]
}
```
I see the `fields` API currently returns `"user.scores": [ 1, 2, 3, 4, 5, 6]` here at the moment, so I will add a test to make the unmapped behaviour doing the same. Or did you think about a different case?

Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java:136]: Interesting variation that cannot happen with mapped fields, but is accepted e.g. with `"dynamic": false`, since we have no mapping and type checks then:
```
{
	"user": [{
		"name": "eve",
		"scores": [1, 2, 3]
	}, {
		"name": "adam",
		"scores": "foo"
	}]
}
```
I think this should return the flattened `"user.scores": [ 1, 2, "foo"]` then, but maybe you disagree on that?

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java:176]: Nice catch, yet another weird edge case. Adding to tests.

File: server/src/test/java/org/elasticsearch/search/fetch/subphase/FieldFetcherTests.java
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/search/fetch/subphase/FieldFetcherTests.java:400]: Adding tests here...

File: server/src/test/java/org/elasticsearch/search/fetch/subphase/FieldFetcherTests.java
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/search/fetch/subphase/FieldFetcherTests.java:643]: Removed some

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldAndFormat.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldAndFormat.java]: After trying Explicit I went with the `null` option here although I don't like it, but it seems the least complex here.

**Comment by cbuescher** [comment]
@jtibshirani thanks a lot for your review so far, I pushed an update that should address your comments.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldAndFormat.java
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldAndFormat.java]: I do agree it's not ideal, but great that it's consistent and keeps the constructors simple!

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java
Comment (jtibshirani) [inline server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java:92]: I just noticed -- maybe we could switch `fieldContexts` to a `LinkedHashMap`? Then we could consult its key set and avoid creating this new set `mappedToExclude`. I could tackle this in a follow-up though, to keep this PR small.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java:141]: Sorry for the confusion -- I didn't spot a behavioral bug, I just found the API choice surprising. The `collectUnmappedList` method returns a list, but `collectUnmapped` adds directly to the document fields map. I was wondering if there needs to be a difference?

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java]: I was thinking something really obvious like `public static final boolean INCLUDE_UNMAPPED_DEFAULT = false`.

**Review by jtibshirani — COMMENTED** [review]
I left some small final comments (accidentally split across two reviews).

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java:92]: Interesting idea, I took a quick look and wasn't sure if we have to consider edge cases where the current list might contain different elements with the same field key though. I agree we should probably do this as a follow up in the interest of getting the basics in this PR in first.

File: server/src/test/java/org/elasticsearch/search/fetch/subphase/FieldFetcherTests.java
Comment (cbuescher) [reply server/src/test/java/org/elasticsearch/search/fetch/subphase/FieldFetcherTests.java:500]: changed that

**Comment by cbuescher** [comment]
@jtibshirani I pushed another update to the branch and think I adressed your additional comments, thanks. Anything else you think needs doing here before merging?

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java
Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java:141]: It looks like this comment is still open. Concretely, could we avoid returning a list from `collectUnmappedList` and just have it add to `documentFields`? This would match the signature for `collectUnmapped`.

**Review by jtibshirani — APPROVED** [review]
This looks good to me. I left one last note about an open comment, but it looks ready to merge after we sort that out.

File: server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java
Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java:141]: Thanks, I understand now. I investigated this and unfortunately it is not as straight-forward as it seems, at least I didn't find an easy way after a bit of searching. I now remember the reason I added the return type was another case I have in the tests. Consider this:
```
Mapping:
{
	"_doc": {
		"properties": {
			"obj": {
				"type": "object",
				"enabled": "false"
			}
		}
	}
}

Doc:
{
	"obj": ["string_value", {
			"a": "b"
		},
		[1, 2, 3]
	]
}
```
I would expect the fields output for `obj` to be `[ "string_value, [1, 2, 3]]` (note the list inside the list. Do do this I need to be able to add to the list collected when on the "obj" level (see  https://github.com/elastic/elasticsearch/pull/65386/files#diff-3c196c8a285307ef760ada7bc5a1df27dce1f4e7160bf75d028833793d171a9aR185). 
At the same time the case above (adding to the same key from two different nested objects should return a single concatenated list) should work. I wasn't able to reconcile both cases without the return type so far.
If you agree with the above expectations on list-inside-list maybe we can leave this investigation for a follow up as well since its an internal anyway?

Comment (jtibshirani) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java:141]: Got it, that is indeed a tricky example. I would actually expect that lists of lists get flattened, as we do for mapped fields. For example, we accept and index documents like this (although maybe we shouldn't):

```
{ "field": [[1, 2, 3]] }
```

The `fields` option will descend into the arrays and return `{ "field": [1, 2, 3] }`. So maybe we can just do the same in your example and return `{ "field": ["string_value", 1, 2, 3] }`.

Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java:141]: Got it, with those expectations I think I can alter the code and test to get around without the return type. I'm not particularly interested in the expected behaviour I mentioned earlier as long as you think flattening is okay here, and since that seems to be the behaviour for mapped types as well I think its okay.

Comment (cbuescher) [reply server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java]: See latest commit and let me know if you are okay with this, otherwise I'd merge and backport to 7.x once CI is green

**Review by jtibshirani — APPROVED** [review]

**Pull request merged**
